communication/mavlink_waypoint_handler.cpp

```diff
@@ -517,1514 +385,285 @@ void Mavlink_waypoint_handler::clear_waypoint_list(Mavlink_waypoint_handler* way
+void Mavlink_waypoint_handler::advance_to_next_waypoint()
 {
-    mav_result_t result;
-
-    if ((waypoint_handler->navigation_.internal_state_ == Navigation::NAV_NAVIGATING) || (waypoint_handler->navigation_.internal_state_ == Navigation::NAV_HOLD_POSITION)
-        || (waypoint_handler->navigation_.internal_state_ == Navigation::NAV_STOP_ON_POSITION) || (waypoint_handler->navigation_.internal_state_ == Navigation::NAV_STOP_THERE))
+    // If the current waypoint index is the last waypoint, go to first waypoint
+    if (current_waypoint_index_ == (waypoint_count_-1))
     {
-        result = MAV_RESULT_ACCEPTED;
-
-        waypoint_handler->navigation_.auto_landing_behavior = Navigation::DESCENT_TO_SMALL_ALTITUDE;
-        waypoint_handler->auto_landing_next_state_ = false;
-
-        waypoint_handler->navigation_.internal_state_ = Navigation::NAV_LANDING;
-
-        //waypoint_handler->navigation_.dubin_state = DUBIN_INIT;
-
-        local_position_t landing_position = waypoint_handler->ins_.position_lf();
-        landing_position[Z] = -5.0f;
-        if (packet->param1 == 1)
-        {
-            print_util_dbg_print("Landing at a given location\r\n");
-
-            landing_position[X] = packet->param5;
-            landing_position[Y] = packet->param6;
-        }
-        else
-        {
-            print_util_dbg_print("Landing on the spot\r\n");
-        }
-
-        if (waypoint_handler->navigation_.navigation_strategy == Navigation::strategy_t::DUBIN)
-        {
-            waypoint_handler->dubin_hold_init(landing_position);
-        }
-        else
-        {
-            waypoint_handler->waypoint_hold_coordinates.position = landing_position;
-            waypoint_handler->waypoint_hold_coordinates.heading  = coord_conventions_get_yaw(waypoint_handler->ahrs_.qe);
-        }
-
-        print_util_dbg_print("Auto-landing procedure initialised.\r\n");
-
-        print_util_dbg_print("Landing at: (");
-        print_util_dbg_print_num(waypoint_handler->waypoint_hold_coordinates.position[X], 10);
-        print_util_dbg_print(", ");
-        print_util_dbg_print_num(waypoint_handler->waypoint_hold_coordinates.position[Y], 10);
-        print_util_dbg_print(", ");
-        print_util_dbg_print_num(waypoint_handler->waypoint_hold_coordinates.position[Z], 10);
-        print_util_dbg_print(", ");
-        print_util_dbg_print_num((int32_t)(waypoint_handler->waypoint_hold_coordinates.heading * 180.0f / 3.14f), 10);
-        print_util_dbg_print(")\r\n");
+        set_current_waypoint_index(0);
     }
-    else
+    else // Update current in both waypoints
     {
-        result = MAV_RESULT_DENIED;
+        set_current_waypoint_index(current_waypoint_index_+1);
     }
-
-    return result;
 }
 
-void Mavlink_waypoint_handler::auto_landing_handler()
+void Mavlink_waypoint_handler::control_time_out_waypoint_msg()
```

Comment:
It is not possible to send and receive waypoints at the same time. When a message for starting to send or receive is read, the autopilot sets the other flag to false.

- If the drone was receiving waypoints and got a message to start sending, it would stop accepting waypoints and do nothing with any that it received. It would then start sending waypoints.
- If the drone was sending waypoints and got a waypoint (not a start receiving message), it would do nothing with it as well.
- If the drone was sending waypoints and got a start receiving message, it would stop sending the waypoints back despite any requests.

One of the changes I made was to switch some variables so that the drone knows how many waypoints have been loaded onto it (before it knew how many were supposed to be there). So if the drone receives waypoints and then has to start sending them, it may end up with an shorter list of waypoints, but it should never attempt to go to a waypoint that has not been received yet.

I suppose another improvement would be to require both of the sending and receiving flags to be false in order to start sending them, that way the drone will not be interrupted when receiving the list.

Thoughts?